File: backend/repository/item.go
package repository

import (
	"context"

	"cloud.google.com/go/firestore"

	"github.com/wheatandcat/Peperomia/backend/domain"
)

// ItemRepository is repository for item
type ItemRepository struct {
}

// NewItemRepository is Create new ItemRepository
func NewItemRepository() domain.ItemRepository {
	return &ItemRepository{}
}

func getItemDocID(uID string, itemID string) string {
	doc := uID + "_" + itemID

	return doc
}

// Create アイテムを作成する
func (re *ItemRepository) Create(ctx context.Context, f *firestore.Client, i domain.ItemRecord) error {
	iDoc := getItemDocID(i.UID, i.ID)

	_, err := f.Collection("items").Doc(iDoc).Set(ctx, i)

	return err
}

// Update アイテムを更新する
func (re *ItemRepository) Update(ctx context.Context, f *firestore.Client, i domain.ItemRecord) error {
	iDoc := getItemDocID(i.UID, i.ID)

	_, err := f.Collection("items").Doc(iDoc).Set(ctx, i)

	return err
}

// Delete アイテムを削除する
func (re *ItemRepository) Delete(ctx context.Context, f *firestore.Client, i domain.ItemRecord) error {
	iDoc := getItemDocID(i.UID, i.ID)

	_, err := f.Collection("items").Doc(iDoc).Delete(ctx)

	return err
}

// FindByUID ユーザーIDから取得する
func (re *ItemRepository) FindByUID(ctx context.Context, f *firestore.Client, uid string) ([]domain.ItemRecord, error) {
	var items []domain.ItemRecord
	matchItem := f.Collection("items").Where("uid", "==", uid).Documents(ctx)
	docs, err := matchItem.GetAll()
	if err != nil {
		return items, err
	}

	for _, doc := range docs {
		var item domain.ItemRecord
		doc.DataTo(&item)
		items = append(items, item)
	}

	return items, nil
}

// DeleteByUID ユーザーIDから削除する
func (re *ItemRepository) DeleteByUID(ctx context.Context, f *firestore.Client, uid string) error {
	matchItem := f.Collection("items").Where("uid", "==", uid).Documents(ctx)
	docs, err := matchItem.GetAll()
	if err != nil {
		return err
	}

	for _, doc := range docs {
		if _, err := doc.Ref.Delete(ctx); err != nil {
			return err
		}
	}

	return nil
}

File: PeperomiaNative/src/lib/firestore/item.ts
import * as firebase from 'firebase';
import 'firebase/firestore';
import { Item as ItemParam } from '../../domain/item';
import { getUUID } from './';

export type Item = ItemParam & {
  id?: string;
  uid: string;
};

const collectionName = 'items';

export const insert = async (
  db: firebase.firestore.Firestore,
  item: Item
): Promise<boolean> => {
  try {
    item.id = getUUID(collectionName);
    const DocRef = db.collection(collectionName).doc(item.id);

    await DocRef.set(item);
    return true;
  } catch (e) {
    console.log(e);
    return false;
  }
};

export const findByUID = async (
  db: firebase.firestore.Firestore,
  uid: string
): Promise<Item[]> => {
  const qs = await db
    .collection(collectionName)
    .where('uid', '==', uid)
    .get();

  const records = qs.docs.map(elem => {
    return elem.data();
  });

  return records as Item[];
};
